md4c/md4c.c

```diff
@@ -3872,14 +3968,6 @@ md_analyze_inlines(MD_CTX* ctx, const MD_LINE* lines, int n_lines, int table_mod
         return ret;
     }
 
-    /* (3) Links. */
-    md_analyze_marks(ctx, lines, n_lines, 0, ctx->n_marks, _T("[]!"));
-    MD_CHECK(md_resolve_links(ctx, lines, n_lines));
-    BRACKET_OPENERS.head = -1;
-    BRACKET_OPENERS.tail = -1;
-    ctx->unresolved_link_head = -1;
-    ctx->unresolved_link_tail = -1;
```

Comment:
It seems like the issue goes away if I revert this, so that links are analyzed before tables.

Reply:
My mistake. It's still there, but the crash/bad output does not happen all the time, and I did not run md2html with the bad input enough times.

Reply:
No, wait. It actually fixes the issue. I was jumping back and forth between branches. What I meant was: If the links analysis is moved to *after* the table analysis, it fixes the issue. So I will attempt to dig a bit deeper into this. Sorry for the confusion.

Reply:
Update: Analyzing links before tables produces the same crash on the master branch too.

@mity: Can you tell if I did something stupid when moving the link analysis to occur before the table analysis? That would be in `md_analyze_inlines`.

Reply:
As a sidenote, I seem to have found a related issue. This is on the master branch with no changes on top.

----
```
[a]: xyz

[abc]
```
```
<p><a href="xyz">abc</a></p>
```
----
```
[abc]: xyz

[a]
```
```
<p><a href="xyz">a</a></p>
```
----
```
[abc]: ijk
[a]: xyz

[a]
```
```
<p><a href="xyz">a</a></p>
```
----
```
[abb]: ijk
[a]: xyz

[a]
```
```
<p><a href="ijk">a</a></p>
```
----
```
[a]: ijk
[abc]: xyz

[abc]
```
```
<p><a href="xyz">abc</a></p>
```
----
```
[a]: ijk
[abb]: xyz

[abb]
```
```
<p><a href="ijk">abb</a></p>
```

Reply:
> It seems like the issue goes away if I revert this, so that links are analyzed before tables. Will dig a bit deeper into this...

Interesting.

> As a sidenote, I seem to have found a related issue. This is on the master branch with no changes on top.

Good catch. Will take a look at it.

Reply:
I have fixed the link label comparison in master. With that applied, you need another an updated test-case for the crash. This "works" for me:

```
[a|b]: /url

x|x
---|---
[a|b]
```

Reply:
Another clue: the body of `if (is_link)` in `md_resolve_links` is executed twice when the crash/bad output appears, but only once when it runs without errors.

Reply:
Another clue: in `md_process_table_row`, `n` is at some point 0 when md2html is called with the crashing input. This may result in `malloc` returning NULL (which is considered), but it may also result in an address that should not be dereferenced, according to my understanding. The latter case is not taken into consideration it seems.

Reply:
Playing in debugger with it a little bit.

I can see that with this input
```
[a|b]: /url

x|x
---|---
`a|b`
```

`md_is_table_row()` returns `FALSE` for the last line. That is fine because the pipe is hidden inside the code span.

But if it is the link as in
```
[a|b]: /url

x|x
---|---
[a|b]
```

then `md_is_table_row()` returns `TRUE`. That is wrong. Maybe this comes from some processing of the link contents because unlike for code spans, links may contain nested inline syntax elements so their contents is recursively re-analyzed from `md_resolve_links()` similarly as a normal paragraph. But that's just pure speculation at this moment.

Anyway, disabling this link contents processing when called in the context of `md_is_table_row()` is imho desirable even if it would not cause the issue: it is clearly an unneeded work. In this context we just need to see whether there is any pipe which marks cell boundary; i.e. any pipe which is _not_ inside any code span or link or anything else with higher priority.

Reply:
Just seeing your two most recent comments. I would say those are quite likely just consequences of `md_is_table_row()` getting bad answer in the 1st place.

Reply:
I am not sure if I understand correctly, but I tested disabling the link analyzing and resolving in `md_analyze_inline` when it is called from `md_is_table_row`, but it does not solve the issue.

Reply:
I see, maybe the culprit is elsewhere in `md_is_table_row()`. But it indeed **should** return `FALSE`. The pipe is inside the link body, links have higher priority and therefore it is not usable as a table cell boundary.

Reply:
`md_is_link_reference` is returning false when it should be true in `md_resolve_links`. It is the call that is commented with `/* Might be collapsed reference link */`.

Reply:
Ouch. Ouch. Ouch. That makes it all clear to me. And it points to an aspect which may make all the idea of links having higher priority then tables very problematic. We have a very ugly catch-22 here.

The problem is that `md_is_table_row()` is called when we are still determining whether the line belongs to a table block. To detect where the table ends and another block starts. I.e. during the very 1st pass for block analysis. Yes, `md_is_table_row()` called some inline analysis of the given line to resolve the higher precedence of some stuff like code spans, escaped pipes over the pipes used so the code spans in the table are possible and it also elegantly solves the case of escaped pipes.

The issue with links is that they can only be really resolved when complete block analysis is over and we know **all** link reference definitions in the document. (Because of this, the hash-table is currently only populated after we know all of them and that's why we hit the issue even if the ref. def. is before.)

Reply:
Interesting, I have just tried to do some experiments with github's Markdown. They clearly do not stop the table when there is a line without pipe.

```
A | B
--- | ---
no-pipe-here-but-still-a-table-row
```

renders into 

A | B
--- | ---
no-pipe-here-but-still-a-table-row

Hence, they likely do not need any line analysis at that early time. In that case I think we can change table parsing in a similar way. Improving the compatibility with them is a good thing on its own.

If I am right (and I believe I am) it should then fix the crash as a side effect. Actually it might even simplify quite a lot some things around the table parsing, maybe even remove the `md_is_table_row()` completely.

 I'll try to cook something (in another branch because it is quite independent from this PR.)

Reply:
That sounds like a clever workaround if it works. Looking forward to seeing if it works.